pmlc/conversion/comp/gpu_to_comp.cc

```diff
@@ -0,0 +1,324 @@
+};
+
+struct ReuseAllocatedMemory : public mlir::OpRewritePattern<comp::Dealloc> {
+  /// Rewrites chain of device memory dealloc and alloc into reuse of previously
+  /// allocated memory. Does not cross block boundaries.
+  using mlir::OpRewritePattern<comp::Dealloc>::OpRewritePattern;
+
+  mlir::LogicalResult
+  matchAndRewrite(comp::Dealloc op,
+                  mlir::PatternRewriter &rewriter) const override;
+};
+
+} // namespace
+
+mlir::LogicalResult
+RewriteLaunchFunc::matchAndRewrite(gpu::LaunchFuncOp op,
+                                   mlir::PatternRewriter &rewriter) const {
+  // Replace gpu.launch_func %indexes..., %args... with:
+  // %env = comp.create_execenv
+  // %new_args = (comp.alloc %env %args)...
+  // %ev = comp.schedule_func %env {
+  //   gpu.launch_func %indexes..., %new_args...
+  // }
+  // %wr_events = comp.schedule_read %new_args to %args on %env wait for %ev
+  // %final_ev = comp.group_events %wr_events...
+  // comp.wait %final_ev
+  // comp.dealloc %new_args...
+  // comp.destroy_execenv %env
+
+  auto loc = op.getLoc();
+  auto context = rewriter.getContext();
+  auto runtime = execEnvType.getRuntime();
+
+  // Create execution environment
+  // TODO How to control what execenv gets created?
+  auto execEnvOp = rewriter.create<comp::CreateExecEnv>(loc, execEnvType);
+  auto execEnv = execEnvOp.getResult();
+  auto eventType = rewriter.getType<comp::EventType>(runtime);
+
+  // Allocate memory on device
+  std::vector<mlir::Value> newOperands;
+  std::vector<mlir::Value> allocOriginal;
+  std::vector<mlir::Value> allocNew;
+  newOperands.reserve(op.getNumKernelOperands());
+  for (auto &operand : op.getOperands()) {
+    if (auto memRefType =
+            operand.getType().dyn_cast_or_null<mlir::MemRefType>()) {
+      mlir::MemRefType newMemRefType =
+          mlir::MemRefType::Builder(memRefType)
+              .setMemorySpace(execEnvType.getDefaultMemorySpace());
+      auto allocOp =
+          rewriter.create<comp::Alloc>(loc, newMemRefType, execEnv, operand);
+      auto newMemRef = allocOp.getResult();
+      newOperands.push_back(newMemRef);
+      allocOriginal.push_back(operand);
+      allocNew.push_back(newMemRef);
+    } else {
+      newOperands.push_back(operand);
+    }
+  }
+  // Update operands for launch
+  op.getOperation()->setOperands(newOperands);
+  mlir::SymbolRefAttr kernel = op.kernel();
+
+  // Move original op into schedule func
+  auto scheduleFuncOp = rewriter.create<comp::ScheduleFunc>(
+      loc, eventType, execEnv, mlir::Value());
+  auto postScheduleIP = rewriter.saveInsertionPoint();
+  auto newBlock =
+      rewriter.createBlock(&scheduleFuncOp.body(), scheduleFuncOp.body().end());
+  // TODO Is there a better way to move operation into block insted of
+  // clone->erase
+  newBlock->push_back(op.clone());
+  rewriter.create<comp::ScheduleEnd>(loc);
+  // TODO Isn't there RAII-idiomatic way to save/restore insertion point
```

Comment:
Yes, see the `PatternRewriter::InsertionGuard`

Reply:
Thank you, I'll use that